import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/StorageCommitmentService.java

```diff
@@ -0,0 +1,242 @@
+package com.google.cloud.healthcare.imaging.dicomadapter;
+
+import com.google.cloud.healthcare.IDicomWebClient;
+import com.google.cloud.healthcare.IDicomWebClient.DicomWebException;
+import com.google.cloud.healthcare.imaging.dicomadapter.AetDictionary.Aet;
+import com.google.cloud.healthcare.imaging.dicomadapter.monitoring.Event;
+import com.google.cloud.healthcare.imaging.dicomadapter.monitoring.MonitoringService;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import org.dcm4che3.data.Attributes;
+import org.dcm4che3.data.Sequence;
+import org.dcm4che3.data.Tag;
+import org.dcm4che3.data.UID;
+import org.dcm4che3.data.VR;
+import org.dcm4che3.net.ApplicationEntity;
+import org.dcm4che3.net.Association;
+import org.dcm4che3.net.Commands;
+import org.dcm4che3.net.Dimse;
+import org.dcm4che3.net.FutureDimseRSP;
+import org.dcm4che3.net.Status;
+import org.dcm4che3.net.pdu.AAssociateRQ;
+import org.dcm4che3.net.pdu.PresentationContext;
+import org.dcm4che3.net.pdu.RoleSelection;
+import org.dcm4che3.net.service.AbstractDicomService;
+import org.dcm4che3.net.service.DicomServiceException;
+import org.json.JSONArray;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class StorageCommitmentService extends AbstractDicomService {
+
+  private static final int EVENT_ID_ALL_SUCCESS = 1;
+  private static final int EVENT_ID_FAILURES_PRESENT = 2;
+  private static Logger log = LoggerFactory.getLogger(StorageCommitmentService.class);
+  private final IDicomWebClient dicomWebClient;
+  private final AetDictionary aets;
+
+  StorageCommitmentService(IDicomWebClient dicomWebClient, AetDictionary aets) {
+    super(UID.StorageCommitmentPushModelSOPClass);
+    this.dicomWebClient = dicomWebClient;
+    this.aets = aets;
+  }
+
+  @Override
+  protected void onDimseRQ(Association as, PresentationContext pc, Dimse dimse, Attributes cmd,
+      Attributes data) throws IOException {
+    if (dimse != Dimse.N_ACTION_RQ ||
+        !cmd.getString(Tag.RequestedSOPClassUID).equals(UID.StorageCommitmentPushModelSOPClass) ||
+        !cmd.getString(Tag.RequestedSOPInstanceUID)
+            .equals(UID.StorageCommitmentPushModelSOPInstance)) {
+      throw new DicomServiceException(Status.UnrecognizedOperation);
+    }
+
+    MonitoringService.addEvent(Event.COMMITMENT_REQUEST);
+
+    Aet remoteAet = aets.getAet(as.getRemoteAET());
+    if (remoteAet == null) {
+      MonitoringService.addEvent(Event.COMMITMENT_ERROR);
+
+      Attributes rspAttrs = new Attributes();
+      rspAttrs.setString(Tag.ErrorComment, VR.LO, "Unknown AET: " + as.getRemoteAET());
+      as.writeDimseRSP(pc, Commands.mkNActionRSP(cmd, Status.ProcessingFailure), rspAttrs);
+      return;
+    }
+
+    CommitmentReportTask task = new CommitmentReportTask(as.getApplicationEntity(),
+        data, remoteAet);
+    as.getApplicationEntity().getDevice().execute(task);
+    as.writeDimseRSP(pc, Commands.mkNActionRSP(cmd, Status.Success));
```

Comment:
There are 3 ways N-Event-Report report can be delivered. 
- on same association
- on next association initiated by SCU
- on separate association initiated by SCP

Only last option is guaranteed to be supported by SCU, and in this case n-action-rsp means that SCP has accepted the commitment request (and will reply later).

(only https://dicomiseasy.blogspot.com/2012/06/storage-commitment.html states this clearly, actual standard http://dicom.nema.org/medical/dicom/current/output/chtml/part04/sect_J.3.3.html#sect_J.3.3.1.2 is worded more vaguely, but seems to match this interpretation)

Specifically on trying to use same association:
"The SCP may attempt to issue the N-EVENT-REPORT on the same Association, but this operation may fail because the SCU is free to release at any time the Association on which it sent the N-ACTION-Request. "

Reply:
Yeah I also noticed this in the standard, so we're ok to just support option 3. It doesn't really make much of a difference in practice since the task won't complete instantly, but I'd still prefer to send the action resp before we enqueue the event task, that way we know for sure we will send the action response first every time. So can we switch the order of these bottom two lines?

Reply:
Wouldn't it be better to surround whole onDimseRQ with try ... catch (RuntimeException e) {send error response}?
Success response on failing to queue CommitmentReportTask would be misleading.